Proposal for network address translation support

garethsb opened the pull request:
This PR suggests adding a "Simple NAT policy" resource type.

As I understand it, GET and PATCH could be supported for this resource type at:

**/network-devices/{networkDeviceId}/nat/{policyId}**

and/or

**/network-devices/{networkDeviceId}/interfaces/{portId}/nat/{policyId}**

A number of open questions related to this:

* [x] 1. Presumably, POST would be supported at the same URLs ending at the **/nat/** parent resource?
* [x] 2. If so, would the `"id"` property (to be used as `{policyId}` in the URLs) be allocated by the client (Broadcast Controller) or the Network Control API?
* [x] 3. Should GET also be supported at that URL to get a (paged) collection of NAT policies?
* [x] 4. Could a top-level collection of policies be offered as well/instead, making it the Network Controller's job to apply the policy at the correct device (switch) or interface.
* [x] 5. Would GET also be supported at the parent resources like **/network-devices/{networkDeviceId}/interfaces/{portId}** or **/network-devices/{networkDeviceId}/interfaces/**? This information is available in the network device itself, so probably not??

vsachin33 committed:
Strawman proposal for network address translation support

garethsb committed:
Fix constraints on port numbers

garethsb committed:
Fix property naming.

garethsb committed:
Fix whitespace

garethsb committed:
Make ordering of "required" vs. "properties" consistent with other schemas

garethsb committed:
Rename example files
(though without e.g. network-device prefix since there are several possible base routes)

garethsb-sony commented:
The current schema seems to invite factoring out a small schema that could be used to describe both a `"match"` sub-object and a `"translated"` sub-object.

I.e.
```json
{
  "match": {
     "source_ip": "10.1.2.3"
  },
  "translated": {
    "source_ip": "40.1.2.3"
  }
}
```

However, it's not clear to me whether ultimately the `"match_*"` properties should support e.g. wildcards, and the `"translated_*"` ones wouldn't, in which case the same schema could not be used?

garethsb-sony commented:
My action from today's meeting was (a) to update the JSON Schemas as identified above, and (b) to add the appropriate endpoints with the appropriate HTTP verbs (GET, POST, DELETE, not sure about PATCH) to the RAML.

We also discussed that in addition to the Network Device scope and the interface scope, we could add a 'global' scope for NAT policies, i.e. at the API top-level, **/nat/{policyId}**.

The narrowest scoped policy would apply in the event of a conflict between policies. (What about between policies at the same scope? How is that handled by network devices now? Implementation-specific, or should some ordering be apparent?)

Each of these NAT scopes would be optionally implemented by Network Controllers, and could independently respond 501, depending on their underlying capability.

A thought post-meeting... would one of the following names rather that **/nat/** be more in keeping with the existing API resource naming?
* **/address-translation(s)/**
* **/address-translation-policies/**
* **/network-address-translation(s)/**
* **/network-address-translation-policies/**

vsachin33 commented:
My vote is for /network-address-translations/

 

..Sachin

 

 

From: Gareth Sylvester-Bradley <notifications@github.com>
Reply-To: AMWA-TV/nmos-network-control <reply@reply.github.com>
Date: Monday, January 27, 2020 at 11:39 AM
To: AMWA-TV/nmos-network-control <nmos-network-control@noreply.github.com>
Cc: Subscribed <subscribed@noreply.github.com>
Subject: Re: [AMWA-TV/nmos-network-control] Proposal for network address translation support (#50)

 

My action from today's meeting was (a) to update the JSON Schemas as identified above, and (b) to add the appropriate endpoints with the appropriate HTTP verbs (GET, POST, DELETE, not sure about PATCH) to the RAML.

We also discussed that in addition to the Network Device scope and the interface scope, we could add a 'global' scope for NAT policies, i.e. at the API top-level, /nat/{policyId}.

The narrowest scoped policy would apply in the event of a conflict between policies. (What about between policies at the same scope? How is that handled by network devices now? Implementation-specific, or should some ordering be apparent?)

Each of these NAT scopes would be optionally implemented by Network Controllers, and could independently respond 501, depending on their underlying capability.

A thought post-meeting... would one of the following names rather that /nat/ be more in keeping with the existing API resource naming?
/address-translation(s)/
/address-translation-policies/
/network-address-translation(s)/
/network-address-translation-policies/
—
You are receiving this because you are subscribed to this thread.
Reply to this email directly, view it on GitHub, or unsubscribe.

garethsb-sony commented:
@vsachin33 Question: Did you want to allow IPv4/IPv6 prefix matching/translation as well as matching of individual addresses?

 The JSON Schema `ipv4` and `ipv6` formats currently used in the schemas can represent only full addresses, i.e. an `ipv4` format string must be a "dotted-quad" as per [RFC 2673, section 3.2](https://tools.ietf.org/html/rfc2673#section-3.2), without a `/` and prefix length.

vsachin33 commented:
Would be nice to have IPv6 support though it may not be used by any vendors right away, regards ..Sachin

 

From: Gareth Sylvester-Bradley <notifications@github.com>
Reply-To: AMWA-TV/nmos-network-control <reply@reply.github.com>
Date: Tuesday, January 28, 2020 at 3:52 AM
To: AMWA-TV/nmos-network-control <nmos-network-control@noreply.github.com>
Cc: vsachin33 <vsachin33@gmail.com>, Mention <mention@noreply.github.com>
Subject: Re: [AMWA-TV/nmos-network-control] Proposal for network address translation support (#50)

 

@vsachin33 Question: Did you want to allow IPv4/IPv6 prefix matching/translation as well as matching of individual addresses?

The JSON Schema ipv4 and ipv6 formats currently used in the schemas can represent only full addresses, i.e. an ipv4 format string must be a "dotted-quad" as per RFC 2673, section 3.2, without a / and prefix length.

—
You are receiving this because you were mentioned.
Reply to this email directly, view it on GitHub, or unsubscribe.

garethsb committed:
Proposal to refactor NAT policy schema

garethsb-sony commented:
> My action from today's meeting was (a) to update the JSON Schemas as identified above

This proposal is this commit on a sub-branch: https://github.com/AMWA-TV/nmos-network-control/commit/21afafd3ed6625439eaeb7c619d5bf3d31ced603

garethsb committed:
Add NAT resource list GET response schema and example

garethsb committed:
Add network-address-translations resource (also to the base 'child resources' response); this could be a required resource in v1.1, permitted to always respond 501 Not Implemented

garethsb committed:
Fix NAT 'description' property to 'label' for consistency with other resources

garethsb-sony commented:
The commit 6cbbbe24d33e62cd79ebf1c51e65910afe7c7ed6 adds the top-level **/network-address-translations** resource for global NAT policies. In order to avoid three repeated sections in the RAML to add Network Device-specific and interface-specific NAT policies, I need to check how best to extract this as a reusable chunk, perhaps using the 'traits' mechanism.

Doing this raised one new question: should PATCH for any attributes be supported on these resources?

garethsb-sony commented:
In order to avoid significant repetition, I have experimented with using `!include` to allow the NAT API resources to be defined in a separate RAML file. This proposal is on a Git sub-branch **nat-scopes**: currently a single commit ebd661e87358b8400e06d4020f208ff6214f971e.

garethsb-sony commented:
As mentioned in the initial comment, the proposed URL for interface NAT scope has been:

**/network-devices/{networkDeviceId}/interfaces/{portId}/nat/{policyId}**

My concern here is that the `interface.portId` attributes in the Network Devices are specified to be the `ifAlias` of the interface. While this is used as a stable 'handle' to identify an interface, its string format is not well constrained. Our examples include "Ethernet 1/2". If used in the URL, the string would have to be URL encoded as `Ethernet%201%2F2`, especially to ensure that the '/' is not treated as signifying hierarchy. How many server and client implementations are going to struggle with that? (See e.g. http://httpd.apache.org/docs/current/mod/core.html#allowencodedslashes)

If we don't like using `portId`, we're going to need something else, and only an array index or adding a UUID property to each `interface` spring to mind right now.

NEOAdvancedTechnology commented:
RFC 8343 "A YANG Data Model for Interface Management" has a leaf "name" of type "string".   Forward slash is not a reserved character in XML, so NETCONF doesn't have a problem with "Ethernet 1/2".

However RFC 8343 also has a mandatory leaf "if-index" of type int32 {range "1..2147483647"}.

garethsb-sony commented:
> (What about between policies at the same scope? How is that handled by network devices now? Implementation-specific, or should some ordering be apparent?)

**Do IS-06 NAT policies need an explicit priority?** (Like OpenFlow rule priority?)

**And/or can conflicting policies be identified at the point that they are PUT and an error response, 400 Bad Request or 409 Conflict, be generated?**

garethsb-sony commented:
> Doing this raised one new question: should PATCH for any attributes be supported on these resources?

We identified a specific use-case: removing the NAT for one receiver endpoint.

This is potentially even more complicated because the JSON patch algorithm used commonly in NMOS specs doesn't have a way to describe removing one or more specific elements from even a JSON array `["foo", "bar", "baz"]`.

This is why IS-06 **/network-flows/{networkFlowId}** supports PATCH but not for `"receiver_endpoint_ids"` (https://github.com/AMWA-TV/nmos-network-control/blob/v1.0.x/APIs/NetworkControlAPI.raml#L308-L310) and instead has the complicated arrangement of a child resource **/receivers** with POST and **/receivers/{receiverId}** supporting DELETE.

**Do we need to adopt the same approach here for NAT policies?**

garethsb-sony commented:
(This point has been made on Slack) 

Are these NAT policies designed only for UDP, or also TCP?

*If so, didn't every policy ought to include a protocol identifier (`"UDP"`, `"TCP"`, or maybe the protocol number from the IP header)?*

garethsb-sony commented:
RAML/schema review:
* Need to add some 501 responses in NAT.raml
* Need to check the Testing Tool handles the `!include` directive (I've already checked raml2html handles it)

garethsb-sony commented:
> If we don't like using `portId`, we're going to need something else, and only an array index or adding a UUID property to each interface spring to mind right now.

The likely problem with array index is that when interfaces are added or split (for example 100Gb/s into 4 x 25Gb/s) that seems likely to change the ordering of the array. Using `ifAlias` as the id seems least worst, but will need implementers of both API and clients to be VERY CAREFUL. Let's discuss on next call.

garethsb-sony commented:
5. Would GET also be supported at the parent resources like **/network-devices/{networkDeviceId}/interfaces/{portId}** or **/network-devices/{networkDeviceId}/interfaces/**? This information is available in the network device itself, so probably not??

   Yes, this should provide the usual 'child resources' response used in all NMOS specs.

garethsb-sony commented:
One issue with multiple locations in the API for **network-address-translations** is that it makes it harder (requires more requests) for a client to discover the applicable NAT policies.

An alternative implementation would be to have all NAT policies under the top-level **/network-address-translations** resource, and add e.g. a `scope` property to the JSON objects, which could be e.g.
* `""` or omitted (to indicate global scope)
* `"/network-devices/{networkDeviceId}"` (or just `"{networkDeviceId}"`)
* `"/network-devices/{networkDeviceId}/interfaces/{portId}"` (or just `"{networkDeviceId}/{portId}"`

This would allow avoiding the awkward use of `{portId}` in an URL path too!

Thoughts?

garethsb committed:
Add 501 to all resource requests under /network-address-translations per Zoom 30-Jan-2020 review comment

garethsb-sony commented:
>  9. Did you want to allow IPv4/IPv6 prefix matching/translation as well as matching of individual addresses?

Several participants have now expressed the opinion that in a first version of this functionality, only exact matches are necessary, no ranges or wildcards on either IP addresses or port numbers. So let's stick to that unless a compelling use case is demonstrated.

garethsb committed:
Merge pull request #57 from AMWA-TV/nat-tuple

Proposal to refactor NAT policy schema --> PR #50

NEOAdvancedTechnology commented:
It is the consensus of the group that (at least for now) this API should map an individual IP address to another single individual IP address.  In the future, we can consider masks/wildcards.

garethsb-sony commented:
As discussed in today's call, asuming we move to only having a top-level **/network-address-translations** collection, we still want to be able to clearly indicate the scope of each NAT policy and to be able to GET:

* only those policies which apply globally
* those policies that may apply to every interface on a network device (whether defined with global scope or network-device scope)
* those policies that apply to a specific interface

Since basic query syntax only supports exact text match, and uses an implicit 'and' operator when multiple query parameters are specified, I believe this simply isn't possible to achieve in a single GET query.

But at least we should design the data model so that it takes at most three non-overlapping GET requests (one per scope!) using basic query syntax, and that a single RQL query could accomplish the same thing if we were to add support for RQL.

Something like:

**/network-address-translations?scope=** (only global policies) assuming we actually require `scope` to be present and a string, with the empty string meaning globak
**/network-address-translations?scope={networkDeviceId}** (only policies applied to that network device)
**/network-address-translations?scope={networkDeviceId}/{portId}** (only policies applied to that interface of that network device)

In RQL that would be:

**/network-address-translations?query.rql=or(eq(scope,),eq(scope,{networkDeviceId}),eq(scope,{networkDeviceId}/{portId}))** (all of the above in a single query)

The alternative approach I outlined, using separate, optional or `null`-supporting `scope.device_id` and `scope.port_id` properties actually seems to be slightly more awkward to use in both basic query syntax and RQL...

@andrewbonney What do you think? Can we do better while sticking to familiar API idioms?

One thing the above doesn't allow is to GET all the NAT policies that are applied at network device or interface scope, rather than for a specific network device or interface. For that RQL `matches` would be necessary. But I'm not sure this is an important use case.

garethsb committed:
Add "translation_scope"

garethsb committed:
Reflect merge of pull request #57 (commit 3af9a3981a96fe6829ad245b87e4ca96a4850430)

garethsb-sony commented:
> > Doing this raised one new question: should PATCH for any attributes be supported on these resources?

* [ ] 17. FWIW, probably we should allow direct PATCH of the NAT policy `"label"` if nothing else?

> We identified a specific use-case: removing the NAT for one receiver endpoint.
> 
> **/network-flows/{networkFlowId}** [...] has the complicated arrangement of a child resource **/receivers** with POST and **/receivers/{receiverId}** supporting DELETE.
> 
> 14. Do we need to adopt the same approach here for NAT policies?

Before I make the additions to the RAML and schemas for this, we would like another discussion to understand the overlap between/need for both the proposed NAT network device interface scope and specifying `"receiver_endpoint_ids"`.

garethsb committed:
Fix examples in line with 45e02e16c7ad69975a9117a6d750f05118efb98b

garethsb committed:
Merge remote-tracking branch 'origin/nat' into nat-scopes-2

garethsb committed:
Add translation_scope to the examples

garethsb committed:
Add PATCH support for consistency with other resource types, but initially only for label rather than any parameters that affect the meaning of the NAT policy

garethsb committed:
Merge remote-tracking branch 'origin/nat' into nat-scopes-2

garethsb committed:
Merge pull request #61 from AMWA-TV/nat-scopes-2

NAT scopes take 2 --> PR #50

garethsb-sony commented:
> > 14. Do we need to adopt the same approach here for NAT policies?
>
> Before I make the additions to the RAML and schemas for this, we would like another discussion to understand the overlap between/need for both the proposed NAT network device interface scope and specifying `"receiver_endpoint_ids"`.

@vsachin33: The intent of interface support is for scoping purpose. If you know certain endpoints who need to receive translated traffic then you just specify them w/o worrying about to which interface they are connected. Interface support was added to either limit translation on the device or support multiple translation policies on the same device for given (S,G)

@garethsb-sony: Is there is a one-to-one mapping between targeting an Endpoint and scoping the NAT policy to the switch interface it is connected to?

@vsachin33: I believe this would vary vendor to vendor based on their support. Say if you apply NAT policy at virtual layer 3 interface (SVI) – VLAN 100 & network 172.22.31.0/24 and have endpoints connected to different physical ports on the switch eth1/1 & eth1/2. You may choose to apply policy only to eth1/1 by including that receiver endpoint and eth1/2 receives traffic w/o translation. So there won’t be 1:1 mapping between interface and endpoint, Having said that vendor would need to have capabilities to support NAT at layer 2 ports.

@garethsb-sony: OK, I will add the **/receivers** child resource that will allow adding receiver endpoints to a NAT policy via POST and removing an existing receiver endpoint via DELETE **/receivers/{receiverId}**.

garethsb committed:
Add the /network-address-translations/{policyId}/receivers[/{receiverEndpointId}] resources to enable modification of the associated receiver Endpoints

garethsb committed:
Fix typo (thank you, @peterbrightwell!)

garethsb-sony commented:
Comments from Cisco via @vsachin33:

>  7. The narrowest scoped policy would apply in the event of a conflict between policies. (What about between policies at the same scope? How is that handled by network devices now? Implementation-specific, or should some ordering be apparent?)

*	Egress NAT - If you have two rules to say translate S1,G1 -> S2,G2 and S1,G1 -> S3,G3 applied on outgoing interface say Ethernet1/1 then packet will be translated twice in accordance with both the rules. It’s possible that there are two receivers behind that egress interface and one is doing IGMP join to G2 and other to G3. Hence there is no notion of priority for NAT rules.

garethsb-sony commented:
> 15. Are these NAT policies designed only for UDP, or also TCP?
>     * If so, didn't every policy ought to include a protocol identifier (`"UDP"`, `"TCP"`, or maybe the protocol number from the IP header)?

**Jian-Rong Chen (Sony)**

The original objective of NAT proposal was trying to solve a specific legacy device and ST2110 over WAN problem. It is very much concerns with media flows, they are all multicast UDP packets. Restrict the current proposal to UDP solves these problems quite nicely.

TCP is much more complicated. It is unicast only, right now no broadcast controller can handle unicast flow control. It is also bi-directional, to make it work, the broadcast controller has to do a lot more work.

I suggest that the NAT policy will check the protocol number in the IP header is UDP, and ignore the policy If it does not match. This could serve as a kind of error detection, since sending TCP packets using multicast address is itself an error anyway.

garethsb-sony commented:
>17. Should we allow PATCH of the NAT policy meaning (match/translated attributes)? I'm thinking not in first version at least?

**Jian-Rong Chen (Sony)** I couldn’t think of any use case of PATCH of a NAT policy. Maybe a receiver can switch between two streams from the same source node by modifying NAT without change of forwarding tables?

@garethsb-sony OK, keeping as-is for now (only `"label"` is directly PATCHable, plus `"receiver_endpoint_ids"` via POST **/receivers** and DELETE **/receivers/{receiverEndpointId}**).

garethsb committed:
Merge remote-tracking branch 'origin/v1.0.x' into nat

garethsb committed:
Remove "translation_scope" to reflect proposal to make the NAT policies "intent"-based, and up to the Network Controller and the Network Devices to apply them appropriately

garethsb committed:
Merge pull request #70 from AMWA-TV/nat-intent

"Intent"-based NAT policies don't need "translation_scope"

garethsb committed:
Clarification around provision of persistent, unique identifiers

garethsb committed:
Merge remote-tracking branch 'origin/v1.0.x' into nat

garethsb committed:
Merge remote-tracking branch 'origin/garethsb-id-docs' into nat

garethsb committed:
Fix reference to query parameters in the single NAT policy endpoint

garethsb committed:
Add the basic documentation for NAT policies.
Examples and diagrams to follow.

garethsb committed:
Update examples to reflect merge of pull request #57 (commit 3af9a3981a96fe6829ad245b87e4ca96a4850430)

garethsb committed:
Add examples for NAT to docs
(Each NATn.svg has been generated from the Visio diagram NAT.vsdx)

garethsb committed:
Use schema property "dependencies" to prohibit wildcard matches on IP address - if port is included in the "match" constraints, the related IP address must also be matched. To prohibit matching all packets, make the "match" tuple required.

NEOAdvancedTechnology committed:
Merge pull request #73 from AMWA-TV/nat-tuple-constraints

Use schema property "dependencies" to prohibit wildcard IP address matching

garethsb committed:
Remove reference to translation scope
(cf. commit 597df639f84538699cb59c26ac06f20c0d299b5f merged in commit 125bb94bc69f87e44916462511c92ec1a4fc731c)

rmsporter committed:
Updated NAT diagrams

rmsporter committed:
Updated wording for NAT examples in docs

garethsb committed:
Move "minProperties" constraint back into the NAT tuple definition because in all cases where the tuple is used, at least one member is required for the NAT policy to be useful

rmsporter committed:
Updated NAT policy wording and diagrams for clarity

garethsb committed:
Merge pull request #74 from robp-sony/nat

Updated NAT examples

garethsb-sony commented:
With some updated examples based on the Network Control activity group call this week, this PR is ready for final review and merging into the **v1.1-dev** branch! 🤞

garethsb commented on docs/3.0. Data Model.md:
```diff
@@ -22,18 +23,27 @@ UUIDs are used throughout the data model, and can be generated in a number of wa
+### Network Address Translation policy
+
+The client registering the Network Address Translation policy is responsible for providing a persistent, unique `id`.
+If the network controller does not have a suitable value, a random (version 4) UUID may be used.
```
```suggestion
If the client does not have a suitable value, a random (version 4) UUID may be used.
```

garethsb committed:
Fix copy&paste mistake

garethsb committed:
Wildcard matches are not permitted in the latest NAT specification

garethsb committed:
Update second NAT example to match latest docs

garethsb committed:
Match updated examples, consistent with docs
